theano/sandbox/multinomial.py

```diff
@@ -215,6 +215,146 @@ def make_node(self, pvals, unis, n=1):
+            %(fail)s;
+        }
+
+        pvals_copy = (PyArrayObject*) PyArray_EMPTY(2,
+            PyArray_DIMS(%(pvals)s),
+            PyArray_TYPE(%(pvals)s),
+            0);
+
+        if (!pvals_copy)
+        {
+            PyErr_SetString(PyExc_MemoryError, "failed to alloc pvals_copy");
+            %(fail)s;
+        }
+        PyArray_CopyInto(pvals_copy, %(pvals)s);
+
+        if ((NULL == %(z)s)
+            || ((PyArray_DIMS(%(z)s))[0] != (PyArray_DIMS(%(pvals)s))[0])
+            || ((PyArray_DIMS(%(z)s))[1] != %(n)s)
+        )
+        {
+            Py_XDECREF(%(z)s);
+            npy_intp dims[2];
+            dims[0] = PyArray_DIMS(%(pvals)s)[0];
+            dims[1] = %(n)s;
+            %(z)s = (PyArrayObject*) PyArray_EMPTY(2,
+                dims,
+                %(t)s,
+                -1);
+            if (!%(z)s)
+            {
+                PyErr_SetString(PyExc_MemoryError, "failed to alloc z output");
+                %(fail)s;
+            }
+        }
+
+        { // NESTED SCOPE
+
+        const int nb_multi = PyArray_DIMS(%(pvals)s)[0];
+        const int nb_outcomes = PyArray_DIMS(%(pvals)s)[1];
+        const int n_samples = %(n)s;
+
+        // for (int n = 0; n < nb_multi; ++n)
+        // {
+        //     for (int m = 0; m < nb_outcomes; ++m)
+        //     {
+        //         const dtype_%(pvals)s* pvals_nm = (dtype_%(pvals)s*)PyArray_GETPTR2(%(pvals)s, n, m);
+        //         dtype_%(pvals)s* pvals_copy_nm = (dtype_%(pvals)s*)PyArray_GETPTR2(pvals_copy, n, m);
+        //         *pvals_copy_nm = *pvals_nm;
+        //
+        //         // if (*pvals_copy_nm == *pvals_nm){
+        //         //     printf("OK");
+        //         // }
+        //     }
+        // }
+
+        //
+        // For each multinomial, loop over each possible outcome,
+        // and set selected pval to 0 after being selected
+        //
+        for (int c = 0; c < n_samples; ++c){
+            for (int n = 0; n < nb_multi; ++n)
+            {
+                double cummul = 0.;
+                const dtype_%(unis)s* unis_n = (dtype_%(unis)s*)PyArray_GETPTR1(%(unis)s, c*nb_multi + n);
+                dtype_%(z)s* z_nc = (dtype_%(z)s*)PyArray_GETPTR2(%(z)s, n, c);
+                for (int m = 0; m < nb_outcomes; ++m)
+                {
+                    dtype_%(z)s* z_nc = (dtype_%(z)s*)PyArray_GETPTR2(%(z)s, n, c);
+                    dtype_%(pvals)s* pvals_nm = (dtype_%(pvals)s*)PyArray_GETPTR2(pvals_copy, n, m);
+                    cummul += *pvals_nm;
+                    if (cummul > *unis_n)
+                    {
+                        *z_nc = m;
+                        *pvals_nm = 0.;
+                        // renormalize the nth row of pvals
+                        dtype_%(pvals)s sum = 0.;
+                        dtype_%(pvals)s* pvals_n = (dtype_%(pvals)s*)PyArray_GETPTR2(pvals_copy, n, 0);
+                        for (int k = 0; k < nb_outcomes; ++k)
```

Comment:
This might not be much more efficient but you could avoid summing over pvals again. You already accumulated `cummul` up to `m` so:

```c
dtype_%(pvals)s sum = cummul - pvals_nm;
if ( m + 1 < nb_outcomes ) {
    dtype_%(pvals)s* pvals_n = (dtype_%(pvals)s*)PyArray_GETPTR2(pvals_copy, n, m + 1);
    for (int k = m + 1; k < nb_outcomes; ++k)
    {
        sum = sum + *pvals_n;
        pvals_n++;
    }
}
```